### PR TITLE
Additional detail in boilerplate

### PR DESCRIPTION
**Pull request by triblondon**
* Mutating the client request is a common use case, so receive the client request in `handle_request` as mutable.
* Make an example mutation to the request (and also to avoid rust complaining that the `mut` is unnecessary)
* Add Jake's mechanism for passing runtime errors into the HTTP response
* Add a match arm for checking method against a list of allowed methods
* Some minor spacing and copy changes which (arguably) improve clarity

### CODE REVIEW

File: src/main.rs
Comment (phamann) [inline src/main.rs]: Whilst I like your solution, you can avoid creating the anonymous vector with an OR operator in your guard conditional. Such as:

```
(method, _) if method != Method::GET || method != Method::HEAD => Ok(...)
```
It's not as clean or terse as yours, but avoids the memory assignment. 

Another way would be to use "Or patterns", which are still experimental, such as:
```
(Method::POST | Method::PUT, _) => Ok(...)
```

/ cc @acfoltzer @katie-martin-fastly for suggestions / thoughts

File: src/main.rs
Comment (phamann) [inline src/main.rs:45]: Why have you added the whitespace between each match arm? Have you ran `cargo fmt` on this?

File: src/main.rs
Comment (phamann) [inline src/main.rs:63]:
```suggestion
            let mut resp = Response::new(e.msg.as_bytes());
```

**Review by phamann — CHANGES_REQUESTED** [review]
Some nice additions. A few minor nits and would like either @acfoltzer or @katie-martin-fastly to review before we merge.

File: src/main.rs
Comment (cratelyn) [reply src/main.rs]: Good catch on the point that arbitrarily nested or patterns are [still experimental](https://github.com/rust-lang/rfcs/blob/master/text/2535-or-patterns.md).

I think I would lean towards the first suggestion in your comment above, @phamann, for two reasons:

* Routing _is_ listed as a use case in the or-pattern RFC, but I think introducing the cognitive load of [enabling unstable features](https://doc.rust-lang.org/unstable-book/language-features/or-patterns.html) on top of the existing "_how does Rust work?_" context is asking a lot from users.
* Judging by the [tracking issue](https://github.com/rust-lang/rust/issues/54883), there is also still some work to be done before this feature stabilizes. If there's another way, i.e. using `if foo || bar =>`, then that's probably best for now.

File: src/main.rs
Comment (cratelyn) [reply src/main.rs:45]: I'd agree here that whitespace between match arms isn't especially conventional. We do have (good, useful) comments around each arm that I see _why_ this is worth trying, however. Maybe we can consolidate the comments before/inside-of each arm, into a single comment before each arm? I.e.

```rust
// If the request is a `GET` to the `/` path, send to a named backend. Request handling
// logic could go here. For example, send the request to an origin backend and then cache
// the response for one minute.
(&Method::GET, "/") => { req.send("backend-name", ONE_MINUTE_TTL) }
```

File: src/main.rs
Comment (cratelyn) [inline src/main.rs:57]: :+1: Lovely! I like how you mushed the comments below into a single doc-comment on `main`. It is more readable, and it also gently shows people how to document their Rust code. :tada:

File: src/main.rs
Comment (cratelyn) [inline src/main.rs:64]: This is _very_ much out of scope for this PR, but I want to point something out:

There will likely be situations early-on where people debugging their code will likely be confused about whether a response w/ an internal server error code represents an error thrown _here_, or an error thrown by an error internal to C@E.

I don't have a great answer for what to do about that, but I want to point out a "_this might be very confusing to a CSE in the future_" wrinkle when I see one.

File: src/main.rs
Comment (cratelyn) [inline src/main.rs]: :+1: Nice concise example for mutating the request.

**Review by cratelyn — APPROVED** [review]
The code in `main.rs` needs to be run through `cargo fmt`, but otherwise I think these are all sensible changes to the boilerplate template. I like this :slightly_smiling_face:

File: src/main.rs
Comment (acfoltzer) [reply src/main.rs]: You could also apply the distributive property to get a top-level or-pattern:
```suggestion
        (Method::HEAD, _) | (Method::GET, _) => Ok(
```
I think I'd favor this because pattern matching is preferred in general to `==`.

Comment (acfoltzer) [reply src/main.rs]: ... I just realized I was missing a negation. I would actually structure this differently. Above the current fallthrough case, I would add a `(Method::GET, path) =>` case that gives the 404 response. Then, the fallthrough case must be an unsupported method, so you could return the appropriate error.

Beyond resolving the immediate syntactic questions, this also is more defensive against future changes to the `Method` type.

**Review by acfoltzer — CHANGES_REQUESTED** [review]
This looks good to go once we resolve the questions around the `match`

File: src/main.rs
Comment (triblondon) [reply src/main.rs]: So the reason I'd prefer to keep my proposed code here is that it is more extensible. The inline vector scales better if I were to add more methods, or if I wanted to pull out the list of methods into a constant or a lookup.  Is there a better way to suggest that potential?

File: src/main.rs
Comment (triblondon) [reply src/main.rs:45]: I did not run `cargo fmt`.  The whitespace is intended to help group things logically, but if angers the Rust gods then so be it.

File: src/main.rs
Comment (triblondon) [reply src/main.rs:64]: True, and I really dislike having this code in customer land - it is intended to be a stopgap until we provide error handling by default.

File: src/main.rs
Comment (acfoltzer) [reply src/main.rs]: So, my main objection here is not the use of a list of methods (though I would recommend using a slice `&[Method::HEAD, Method::GET]` rather than a vec to avoid allocation), but that logically this is expressing a fallthrough case, and so in a Rust program I would expect to see it as a fallthrough pattern.

Perhaps this could be transformed to a clause that matches if the method is a member of a list, as opposed to ensuring that the list doesn't contain the method?

File: src/main.rs
Comment (triblondon) [reply src/main.rs:45]: Hmm. I ran rustfmt in VSCode and it didn't remove these blank lines 🤷‍♂

File: src/main.rs
Comment (triblondon) [reply src/main.rs:57]: Thanks!

File: src/main.rs
Comment (triblondon) [reply src/main.rs]: Abstracted this out.  What do you think?

Comment (phamann) [reply src/main.rs]: Whilst I like your new abstraction, I'm confused as it isn't the same logic as before. It seems to be allowing all methods not just a subset i.e. before you were only allowing GET and HEAD.

Comment (triblondon) [reply src/main.rs]: Customer will customise so it doesn't matter that much, but the new impl was inconsistent with the error msg so I updated the message.

Comment (phamann) [reply src/main.rs]: Ok. My only concern with your new list is it allows ALL methods and therefore a strange example, as if a client sends a request with an unknown or malformed method it won't even be forwarded by h2o to xqd and therefore makes your logic redundant. Could we at least reduce the allow list to make a better/more realistic example?

**Review by phamann — APPROVED** [review]
Circling back on this, apologies for the delay. I've just tested the changes locally and very happy with how this is evolving. Thank you for your contributions and patience @triblondon :+1:

**Pull request merged**
